DT-536

will-garces-biblica opened the pull request.

will-garces-biblica committed:
DT-536: Remove Selected column and add Location column

will-garces-biblica committed:
DT-536: Add form for editing TVP settings and update menus

will-garces-biblica committed:
DT-536

- Load saved TVP options into OptionsForm for editing
- Use saved repository settings to setup remote repository instead
  of environment variables
- Verify repository settings before enabling the remote repository
- Add Id, Enabled, and Verified fields to saved repository options
- Enable user control of startup synchronization
- Fix progress form starting position to center of parent forms
- Fix unsafe cross thread control access

mhutton86 commented on CheckManagement/IRepository.cs:
```diff
@@ -18,6 +20,45 @@ namespace TvpMain.CheckManagement
+        /// Warning: this will overwrite an existing admin list.
+        /// </summary>
+        /// <param name="userName">The user name of the admin.</param>
+        public void SetAdmin(string userName);
```
This presumes a singular admin per repo. Would it be better to have `AddAdmin` and `RemoveAdmin`  functions to support multiple admins?

mhutton86 commented on CheckManagement/LocalRepository.cs:
```diff
@@ -26,11 +29,85 @@ public class LocalRepository : IRepository
+        /// <summary>
+        /// A unique identifier for the repository.
+        /// </summary>
+        public Guid Id { get { return new Guid(); } }
```
This appears like it would return a new Guid every single time it's called rather than setting it only once.

will-garces-biblica replied:
Calling `new Guid()` creates a new Guid with the default value. So it will be the same each time it is called. But that is not obvious so I changed it to `default(Guid)` which is a little clearer about the value that will be returned.

mhutton86 commented on CheckManagement/LocalRepository.cs:
```diff
@@ -26,11 +29,85 @@ public class LocalRepository : IRepository
+                return true; 
+            }
+
+            set { }
```
These empty set statements appear unnecessary. C# properties without a set function are considered read-only.

will-garces-biblica replied:
They are necessary because LocalRepository implements IRepository which requires a set clause. IRepository is also implemented by S3Repository which needs the ability to change the Enabled property.

will-garces-biblica replied:
Per your suggestion I have changed the setter on these properties to throw an exception instead of failing silently.

mhutton86 commented on CheckManagement/LocalRepository.cs:
```diff
@@ -26,11 +29,85 @@ public class LocalRepository : IRepository
+        /// <summary>
+        /// A list of admin user names for this repository.
+        /// </summary>
+        public List<string> Admins { get { return new List<string>(); } }
```
This will always return a new list every time it's called. For default values you you can do the following:
`public List<string> Admins { get; set; } = new List<string>(); `
If you don't want to provide the ability to set this list, you can remove the `set` keyword.

will-garces-biblica replied:
I like your suggestion. The syntax looks cleaner. Done.

mhutton86 commented on CheckManagement/LocalRepository.cs:
```diff
@@ -91,5 +168,53 @@ private void VerifyFolderPath()
         }
+
+        // <summary>
+        // Deletes all checks from this repository.
```
Rogue documentation.

will-garces-biblica replied:
Deleted.

mhutton86 commented on CheckManagement/LocalRepository.cs:
```diff
@@ -91,5 +168,53 @@ private void VerifyFolderPath()
+                }
+                return true;
+            }
+            catch (Exception)
```
Instead of failing silently, it may be better to let the exception bubble up or printing a warning message. In case this produces errors that cause defects down the line.

mhutton86 commented on CheckManagement/S3Repository.cs:
```diff
@@ -21,10 +27,143 @@ namespace TvpMain.CheckManagement
     /// </summary>
     public class S3Repository : IRepository
     {
+        private string _accessKey;
```
For consistency, these can be made private properties. Not a dying hill, just a suggestion. This also means you wouldn't have to initialize them in the constructor to empty values.

`private String AccessKey { get; set; } = ""`

mhutton86 commented on CheckManagement/S3Repository.cs:
```diff
@@ -21,10 +27,143 @@ namespace TvpMain.CheckManagement
+            }
+            catch (Exception) 
+            {
+                Verified = false;
```
For the sake of troubleshooting/debugging issues, it's worth logging the exceptions.

mhutton86 commented on CheckManagement/S3Repository.cs:
```diff
@@ -21,10 +27,143 @@ namespace TvpMain.CheckManagement
+                        var stream = Service.GetFileStream(MainConsts.PERMISSIONS_FILE_NAME);
+                        using var reader = new StreamReader(stream);
+                        IEnumerable<string[]> lines = CSVFile.Read(reader);
+                        _admins = lines.ToList().SelectMany(item => item).ToList();
```
This can likely be shortened to `lines.ToList()`

mhutton86 commented on CheckManagement/S3Repository.cs:
```diff
@@ -77,5 +216,83 @@ public void RemoveCheckAndFixItem(string filename)
+        /// <returns>true if the files were successfully deleted. false otherwise.</returns>
+        public bool Clear()
+        {
+            // Deletion of all remote repository files is not currently supported.
```
It's usually better to throw a `NotSupportedException`  or `NotImplementedException`. The UI should catch these exceptions and typically display a "This feature isn't supported" error.

mhutton86 commented on CheckManagement/S3Service.cs:
```diff
@@ -116,5 +124,146 @@ public HttpStatusCode DeleteFile(string filename)
+            }
+            catch (Exception ex)
+            {
+                error = $"Could list files in bucket '{BucketName}'. {ex.Message}";
```
Message is missing the word `not`

will-garces-biblica replied:
Fixed.

mhutton86 commented on CheckManagement/S3Service.cs:
```diff
@@ -116,5 +124,146 @@ public HttpStatusCode DeleteFile(string filename)
+        public static bool ValidateAccessKey(string key)
+        {
+            Regex regex = new Regex(@"^[A-Z0-9]{20}$");
+            if (regex.IsMatch(key))
```
This could be shorted to `return regex.IsMatch(key);`

will-garces-biblica replied:
Thanks for pointing this out. Done.

mhutton86 commented on CheckManagement/S3Service.cs:
```diff
@@ -116,5 +124,146 @@ public HttpStatusCode DeleteFile(string filename)
+        public static bool ValidateSecretKey(string key)
+        {
+            Regex regex = new Regex(@"^[A-Za-z0-9/+=]{40}$");
+            if (regex.IsMatch(key))
```
This could also be shortened like the above.

will-garces-biblica replied:
Done.

mhutton86 commented on Project/OptionsManager.cs:
```diff
@@ -0,0 +1,65 @@
+        /// The full path to the file where TVP options are stored.
+        /// </summary>
+        private static string _optionsFilePath = Path.Combine(Directory.GetCurrentDirectory(),
+                MainConsts.TVP_FOLDER_NAME, MainConsts.CONFIGURATION_FOLDER, OPTIONS_FILE_NAME);
```
Will this work if the plugin has a different plugin name?

will-garces-biblica replied:
It will as long as the plugin folder name does not change. If it does, the folder name must be updated in MainConsts.TVP_FOLDER_NAME.

mhutton86 commented on Project/TvpOptions.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <summary>
+        /// Settings for the remote repository used for sharing checks among TVP users.
+        /// </summary>
+        public List<S3RepositoryOptions> SharedRepositories { get; set; }
```
Why specify a list? if we only appear to support a single repository? It seems that we should commit to supporting either multiple repositories or a single repository.

mhutton86 replied:
Based on our Zoom conversation, please add a comment to specify this is to support the future need of a list of repositories.

will-garces-biblica replied:
I added a comment about supporting multiple remote repositories in the code and the story notes.

will-garces-biblica committed:
DT-536: Fix review comments

mhutton86 committed:
DT-536:

- OptionsManager made to be singleton and support parametized config directory
- CheckEditor update to bring in instance of CheckManager during construction
- CheckManager and RunChecks updated to instantiate and pass around a single instance of OptionsManager
- CheckManagerTests updated to support new OptionsManager class

mhutton86 committed:
DT-536:

- add TestOptions directory for instantianting OptionsManager during tests

mhutton86 committed:
DT-536: Merge and fix change from remote branch

will-garces-biblica committed:
DT-536: Fix refresh button bug and failing CheckManager unit test
